### ISSUE TITLE
Access offline player data

### ISSUE THREAD

**Issue by boformer** [SpongePowered/SpongeAPI#528]
Right now I'm working on a server mail plugin. Of course, players can also send mails to offline players.

I want to check if the player name entered by the player is a registered Minecraft profile name and if the player has played before on the server.

These kind of methods are important for many plugins (e.g. /lastonline <offline-player> or /ban <offline-player>).

I searched the SpongeAPI code for a method to get the ``GameProfile``/``User`` of an offline player, but did not find one.

Proposed methods:
* ``Optional<User> getUser(UUID uuid, boolean playedBefore)``
* ``Optional<User> getUser(String name, boolean playedBefore)``
* ``List<User> getUsers()`` --> Returns list of all users who played before.

The ``getUser`` method should first check if the user exists in the world save files. If it does not exist (user has not played before) and ``playedBefore`` is set to false, the server should contact the Mojang servers to check if the username/UUID exists.

**Comment by MattBDev** [comment]
I do believe this is needed but `getUser` with a string parameter (I am assuming it is accepting a username) wouldn't be able to check the world save files for the player based on their name since mojang keeps no record of their last joined name. Bukkit added their own nbt tag to each player data file but I am not sure if Sponge would be willing to do this.
I also feel that there should be methods without `playedBefore`.

**Comment by modwizcode** [comment]
I disagree with having `playedBefore` anywhere, that's a very strange thing to have. We should maintain a cache for getUser(String username) which will be update by Mojang if no cache and no player online and also updated when the player joins the server. People will be encouraged to use UUID's but we shouldn't make plugins have their own cache.

**Comment by boformer** [comment]
@modwizcode I agree with you that a ``username <-> UUID`` cache is the way to go.

But how would this cache work when a player changes the username?

Example: 
```
modwizcode plays on server.
server saves username in cache.
modwizcode logs out
modwizcode changes name to FakeNotch
Player calls /lastseen modwizcode, which calls getUser("modwizcode")
Player calls /lastseen FakeNotch  , which calls ``getUser("FakeNotch")``
```
What will the method return for ``"modwizcode"``, what for ``"FakeNotch"``?

Maybe we should add 2 different methods for name lookups:
* ``getUserFromCache(String name)`` --> Only searches cached player data 
(``Optional.absent()`` for ``"FakeNotch"``)
* ``getUserFromMojang(String name)`` --> Only contacts Mojang's servers

We can also combine these methods with a boolean parameter.

To make it perfect, there should be a way to get both usernames:

* The username the player had on last login (from cache)
* The username the player currently uses (from Mojang)

**Comment by MattBDev** [comment]
At the moment there are a lot of Bukkit/Spigot plugins that have their own player caches. When you have more and more plugins on your server that manage a player cache in their own way, it is just increasing the amount of file space being used and the memory being used on the server. If Sponge could come up with an effective system that manages a player cache, it would save a lot of memory on servers.

**Comment by modwizcode** [comment]
I know forge provides  a UUID->username cache and exposes the map making
the reverse possible.

>

**Comment by MattBDev** [comment]
Minecraft provides a usercache file but unfortunately on Spigot, server owners have the option to limit the cache size so they could set it to 0 and ruin the cache system.

**Comment by modwizcode** [comment]
But this isn't Spigot... So that doesn't matter at all. This is based on
Forge or Vanilla.

**Comment by TheE** [comment]
I think the approach discussed in the starting comment mixes two completely separated approaches that should better be kept somewhat separated.

On the one hand there is the cache that stores UUID - name bindings. Each UUID is unique, but a name may have multiple UUIDs (e.g. if a player, known to the server as `A`, changes his name to `C` but never comes back to the server and _another_ player, known as `B` changes his name to `A` and joins the server).  Therefore the cache cannot be trusted when resolving names to UUIDs and it comes down to:
````JAVA
public interface UserCache {
    Optional<User> get(UUID uuid);
    List<User> getAll();
}
````

On the other hand there is the resolver that looks up UUIDs from names. This can happen in two ways, either you query online players or you contact Mojang's servers witch results in a blocking call. It comes down to:
````JAVA
public interface UserResolver {
    Optional<User> get(String name);
    List<User> getAll(Iterable<String> names);
}
````
The implementation could be chained so it checks online players before calling Mojang (although the thread safety of the whole operation is another issue, but let us ignore that for a moment). Under all situations I can imagine, all resolved users should go directly into the cache, because otherwise there might not be a way to get the name (think of a user whose UUID was resolved once from a given name, but who was never on the server). Additionally the cache is updated whenever a player logs in but that should be obvious. Both behaviours are not implementation detail but required functions in order to work properly.

On the bottom line, @sk89q has written a nice library for UUID - name conversations, caching etc.: [SquirrelID](https://github.com/sk89q/SquirrelID). To my understanding, it works similar to the system I described above.

**Comment by MattBDev** [comment]
Offline Mode needs to be considered when implemented. https://github.com/sk89q/SquirrelID does not look like it has only support for Mojang UUID's and not offline UUID's.

**Comment by ZephireNZ** [comment]
For the HTTP requests, perhaps use either `Future`, or `ListenableFuture`? That way a plugin can run the request, and choose to block until it's completed, or spin it off into a thread.

**Comment by progwml6** [comment]
@MattBDev for anything that uses mojang's auth system we would most likekly be using authlib from mojang which is used in MC, Mojang's launcher, along with several of the java based modded launchers

**Comment by boformer** [comment]
I can not continue with plugin development without this feature. We should decide how to handle this.

**Comment by Azatik** [comment]
Why not just check the folder with the world. If a player has ever gone to the server, the inventory of the player will be in the folder.

**Comment by simon816** [comment]
@Azatik That's not what plugins should do. There's no specification for where user data is stored in the API, it could even be stored on a remote sql server, plugins don't know.

I think a few methods just need to be added to the `Server` interface
`User createUser(GameProfile profile)`
`User createUser(String username)`
`User createUser(UUID uniqueId)`
`List<UUID> getKnownUserIds()`
Caching should be handled in the implementations, IMO plugins shouldn't be concerned about how the data is fetched.

**Comment by TheE** [comment]
@simon816 I agree that the API should not make any specification where the user data is stored, but I do not see why this enforces to split UUIDs and Users. After all, the sever-cache must store UUIDs and the last-known name associated with this ID. Likewise any remote lookup must resolve names into UUIDs.

I think the addition of two separated services, Cache and Resolver, along the lines of what I said earlier should be sufficient, while both should probably return Futures. In the implementation:
* The cache could be implemented with a simple H2 database or use vanilla's `usercache.json` if possible.
* The resolver could resolve an UUID by checking if an online player has that UUID and, if not, call Mojang's server. Non Minecraft implementations could bundle a different resolver and plugins could even replace Sponge's default resolver with their own.

**Comment by MattBDev** [comment]
@TheE Vanilla Servers do not store the last known name of any player. It only stores the UUID. Bukkit/Spigot created their own nbt tag that did store the last known name of a player. Obviously this can't be applied to the SpongeAPI

**Comment by TheE** [comment]
@MattBDev Vanilla servers keep a locale UUID - name cache in the username.json within the server directory. As far as I know, this cache contains pairings for every player that joined the server and for every player that was requested per a vanilla command.

There might be some internal reasons why we want to use our own cache, but that is defiantly implementation detail and there is no need to worry about it now.

**Comment by simon816** [comment]
Resolved by #580
